function/function.rs

```diff
@@ -4,95 +4,34 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
  */
 
-use answer::Type;
-use encoding::{graph::definition::definition_key::DefinitionKey, value::value_type::ValueType};
-use ir::{
-    pattern::variable_category::{VariableCategory, VariableOptionality},
-    program::function::FunctionIR,
-};
+use encoding::graph::definition::{definition_key::DefinitionKey, function::FunctionDefinition};
+use ir::program::function_signature::FunctionIDTrait;
 
-/// Function represents the user-defined structure:
-/// fun <name>(<args>) -> <return type> { <body> }
-pub struct Function {
-    definition_key: DefinitionKey<'static>,
+use crate::FunctionManagerError;
 
-    // parsed representation
-    name: String,
-    arguments: Vec<FunctionArgument>,
-    return_type: FunctionReturn,
-
-    // pre-compiled arguments, body, return // TODO maybe deferred compilation
-    ir_body: FunctionIR,
-}
-
-struct FunctionArgument {
-    name: String,
-    type_: FunctionValuePrototype,
-}
-
-enum FunctionReturn {
-    Stream(Vec<FunctionValuePrototype>),
-    Single(FunctionValuePrototype),
+pub type SchemaFunction = Function<DefinitionKey<'static>>;
+pub struct Function<FunctionIDType: FunctionIDTrait> {
```

Comment:
You don't need to call types `Type`, just call it `FunctionID`, and itll get used as a type :)